### DISCUSSION TITLE
discriminate in hooks.server if request is from form action vs page first load?

### DISCUSSION THREAD

**Discussion by geko1971** [sveltejs/kit#7905] [Q&A]
I'm wondering if there is a way to discriminate in the server hooks whether an incoming request was issued by an action-triggered page refresh vs a page _initial load_.

Reason is that I run an handle in server hooks that need to validate a certain authorization cookie expiration only when pages first load, but skip that validation for all subsequent refreshes of same page, when those are triggered for instance by form submissions.

It would be great if there could be (in the `$app/navigation` module? In the `event` itself?) a property declaring whether the event is related to page _first load_ or a subsequent reload.

**Comment by david-plugge** [comment]
There´s a header called `x-sveltekit-action` that is set to true when calling with fetch from the client using `use:enhance`

https://github.com/sveltejs/kit/blob/f953c9d810be8b9211ce1fa456d9c96224ec55dc/packages/kit/src/runtime/app/forms.js#L99

**Reply by geko1971** [reply]
Thank you David.  I am using form actions as exported by `+page.server` (not `use:enhance` action), since I need to perform server-side validation/augmentation of POSTed data: unfortunately, `x-sveltekit-action` http header seem not to be included with the requests in the case.

**Comment by teemingc** [comment]
This solution checks for the `content-type` header.
It can detect the request is a form action if it contains `application/x-www-form-urlencoded`.
https://stackblitz.com/edit/sveltejs-kit-template-default-35sq7b?file=src/hooks.server.js

Although, the code in hooks runs on every request, so you'd still have the issue of validation running on each page load. I suggest validating the cookie expiration in a `+layout.server.js` file so it runs only when the layout first mounts.

**Reply by geko1971** [reply]
Hello Tee Ming, thank you for your feedback.

I moved my cookie validation to `+layout.server.js` as suggested, but against expectations I see that the `load: LayoutServerLoad` method at `+layout.server.js` gets called every time each page reloads (even upon form `POST` actions-triggered refreshes).

That is indeed why I was hoping for a method that was instrumented in the framework, to discriminate the first upload from the subsequent ones. Without that instrumentation, it gets really _hacky_. 😞

**Reply by teemingc** [reply]
I think you'll have to implement your own logic to prevent the validation running on full page reloads. Perhaps you could set another cookie to mark the first visit?

Load functions re-run whenever:
* the web page is reloaded (all state is lost except cookies, local storage, etc.)
* the default `POST` form action behaviour calls `invalidateAll()` to re-run all load functions.
* parts of the URL or params is used in the function and it changes.
* child load functions re-run when their parent load re-runs.

For more specifics see the Invalidation section in the docs: https://kit.svelte.dev/docs/load#invalidation

**Reply by stefandevo** [reply]
@s3812497 is there a way to NOT invalidateAll for a POST action?

**Reply by teemingc** [reply]
> @s3812497 is there a way to NOT invalidateAll for a POST action?

Yes, if you provide a custom callback to the `enhance` action 

```svelte
<form
  method="POST"
  use:enhance={({ form, data, action, cancel }) => {
    // `form` is the `<form>` element
    // `data` is its `FormData` object
    // `action` is the URL to which the form is posted
    // `cancel()` will prevent the submission

    return async ({ result }) => {
     // choose what you want to do when result === 'success'
     // this is where it usually invalidates all
    };
  }}
>
```
https://kit.svelte.dev/docs/form-actions#progressive-enhancement-use-enhance


or implement your own submit event listener https://kit.svelte.dev/docs/form-actions#progressive-enhancement-custom-event-listener.

**Reply by stefandevo** [reply]
Ok, but I want to redirect to another page after my Post. In this case it seems there is no way to not invalidateAll(). If I stay on the same page after posting, indeed, I can ignore the invalidateAll() call - but a redirect seems to invalidateAll() nevertheless.

**Reply by teemingc** [reply]
Do you mind sharing an example of how you have the form and action set up?

**Reply by stefandevo** [reply]
I created a small app here:
https://stackblitz.com/edit/sveltejs-kit-template-default-ecme4h?file=README.md

**Reply by teemingc** [reply]
Posted a working example on your other post. Hope it helps! :)
https://github.com/sveltejs/kit/discussions/9467#discussioncomment-5383889
